### PR TITLE
openconnect: fix dependencies to iconv/intl (fixes #14734)

### LINKED ISSUE #14734: openconnect: libxml2: undefined reference to libiconv
Maintainer: @nmav
Compile tested: Turris Omnia, mvebu (cortex-a9), OpenWrt master
Enabled hardening, NLS and other stuff.

Description:

```
SHELL= flock /foo/build/tmp/.openconnect-8.10.tar.gz.flock -c '  	/foo/build/scripts/download.pl "/foo/build/dl" "openconnect-8.10.tar.gz" "30e64c6eca4be47bbf1d61f53dc003c6621213738d4ea7a35e5cf1ac2de9bab1" "" "ftp://ftp.infradead.org/pub/openconnect/"    '
rm -f /foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10/.built
touch /foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10/.built_check
CFLAGS="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -fmacro-prefix-map=/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10=openconnect-8.10 -Wformat -Werror=format-security -fpic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Wl,-z,now -Wl,-z,relro  -I/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/usr/include -I/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/include/fortify -I/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/include " CXXFLAGS="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -fmacro-prefix-map=/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10=openconnect-8.10 -Wformat -Werror=format-security -fpic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Wl,-z,now -Wl,-z,relro  -I/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/usr/include -I/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/include/fortify -I/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/include " LDFLAGS="-L/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/usr/lib -L/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/lib -fpic -specs=/foo/build/include/hardened-ld-pie.specs -znow -zrelro " make -j1 -C /foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10/. AR="arm-openwrt-linux-muslgnueabi-gcc-ar" AS="ccache_cc -c -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -fmacro-prefix-map=/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10=openconnect-8.10 -Wformat -Werror=format-security -fpic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Wl,-z,now -Wl,-z,relro" LD=arm-openwrt-linux-muslgnueabi-ld NM="arm-openwrt-linux-muslgnueabi-gcc-nm" CC="ccache_cc" GCC="ccache_cc" CXX="ccache_cxx" RANLIB="arm-openwrt-linux-muslgnueabi-gcc-ranlib" STRIP=arm-openwrt-linux-muslgnueabi-strip OBJCOPY=arm-openwrt-linux-muslgnueabi-objcopy OBJDUMP=arm-openwrt-linux-muslgnueabi-objdump SIZE=arm-openwrt-linux-muslgnueabi-size CROSS="arm-openwrt-linux-muslgnueabi-" ARCH="arm" ;
make[3]: Entering directory '/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10'
make  all-recursive
make[4]: Entering directory '/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10'
Making all in tests
make[5]: Entering directory '/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10/tests'
/bin/bash ../libtool  --tag=CC   --mode=link ccache_cc  -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -fmacro-prefix-map=/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10=openconnect-8.10 -Wformat -Werror=format-security -fpic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Wl,-z,now -Wl,-z,relro   -L/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/usr/lib -L/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/lib -fpic -specs=/foo/build/include/hardened-ld-pie.specs -znow -zrelro  -o serverhash serverhash.o ../libopenconnect.la -L/foo/build/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/lib -lgnutls   
libtool: link: ccache_cc -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -fmacro-prefix-map=/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10=openconnect-8.10 -Wformat -Werror=format-security -fpic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Wl,-z -Wl,now -Wl,-z -Wl,relro -fpic -specs=/foo/build/include/hardened-ld-pie.specs -znow -zrelro -o serverhash serverhash.o  -L/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/usr/lib -L/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/lib ../.libs/libopenconnect.a -L/foo/build/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/lib -lxml2 -lz -lgnutls
/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/8.4.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: warning: libiconv.so.2, needed by /foo/build/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/lib/libxml2.so, not found (try using -rpath or -rpath-link)
/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/8.4.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: /foo/build/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/lib/libxml2.so: undefined reference to `libiconv'
/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/8.4.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: /foo/build/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/lib/libxml2.so: undefined reference to `libiconv_close'
/foo/build/staging_dir/toolchain-arm_cortex-a9+vfpv3-d16_gcc-8.4.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/8.4.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: /foo/build/staging_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/usr/lib/libxml2.so: undefined reference to `libiconv_open'
collect2: error: ld returned 1 exit status
make[5]: *** [Makefile:768: serverhash] Error 1
make[5]: Leaving directory '/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10/tests'
make[4]: *** [Makefile:1489: all-recursive] Error 1
make[4]: Leaving directory '/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10'
make[3]: *** [Makefile:753: all] Error 2
make[3]: Leaving directory '/foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10'
make[2]: *** [Makefile:89: /foo/build/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/openconnect-8.10/.built] Error 2
make[2]: Leaving directory '/foo/build/feeds/packages/net/openconnect'
time: package/feeds/packages/openconnect/compile#0.40#0.10#0.47
    ERROR: package/feeds/packages/openconnect failed to build.
make[1]: *** [package/Makefile:114: package/feeds/packages/openconnect/compile] Error 1
```

### PR DESCRIPTION
**Pull request by mhei**
Maintainer: @nmav
Compile tested: mxs
Run tested: -

Description:

This fixes the issue raised after d18692c (libxml2: allow building with iconv support).

Signed-off-by: Michael Heimpold <mhei@heimpold.de>

### CODE REVIEW

File: net/openconnect/Makefile [hunk 1]
```diff
@@ -34,7 +35,7 @@ endef
 define Package/openconnect
   SECTION:=net
   CATEGORY:=Network
-  DEPENDS:=+libxml2 +kmod-tun +resolveip +vpnc-scripts +OPENCONNECT_OPENSSL:libopenssl +OPENCONNECT_OPENSSL:p11-kit +OPENCONNECT_OPENSSL:libp11 +OPENCONNECT_GNUTLS:libgnutls +OPENCONNECT_GNUTLS:libtasn1 +OPENCONNECT_STOKEN:libstoken
+  DEPENDS:=+libxml2 +kmod-tun +resolveip +vpnc-scripts +OPENCONNECT_OPENSSL:libopenssl +OPENCONNECT_OPENSSL:p11-kit +OPENCONNECT_OPENSSL:libp11 +OPENCONNECT_GNUTLS:libgnutls +OPENCONNECT_GNUTLS:libtasn1 +OPENCONNECT_STOKEN:libstoken $(ICONV_DEPENDS) $(INTL_DEPENDS)
   TITLE:=OpenConnect VPN client (Cisco AnyConnect and Juniper/Pulse compatible)
```
Comment (nmav) [inline net/openconnect/Makefile:39]: I do not think that this is the right solution. openconnect doesn't depend on iconv, so it shouldn't link to it. Since libxml2 depends in it, it should have done the right thing and added the right dependencies. That's what the rest of the libraries linked with like gnutls and openssl do.

Comment (micmac1) [reply net/openconnect/Makefile:39]: Funnily enough openconnect does actually (try to) use iconv:

```
Disabling Docbook support
Enabled Schematron support
checking iconv.h usability... yes
checking iconv.h presence... yes
checking for iconv.h... yes
checking for iconv... no
checking for iconv in -liconv... yes
checking for iconv declaration...
        extern size_t iconv (iconv_t cd,  char * *inbuf, size_t *inbytesleft, char * *outbuf, size_t *outbytesleft);
Disabling ICU support
```

It seems to use iconv from libc if available. But when you tell it that you want it to use external libiconv/libintl it will:

```
diff --git a/net/openconnect/Makefile b/net/openconnect/Makefile
index 59e81f77a..3fd616227 100644
--- a/net/openconnect/Makefile
+++ b/net/openconnect/Makefile
@@ -26,6 +26,7 @@ PKG_CONFIG_DEPENDS:= \
 PKG_USE_MIPS16:=0
 
 include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/nls.mk
 
 define Package/openconnect/config
        source "$(SOURCE)/Config.in"
@@ -55,6 +56,8 @@ endef
 
 CONFIGURE_ARGS += \
        --disable-shared \
+       --with-libiconv-prefix=$(ICONV_PREFIX) \
+       --with-libintl-prefix=$(INTL_PREFIX) \
        --with-vpnc-script=/lib/netifd/vpnc-script \
        --without-libpcsclite \
        --without-stoken \
```

readelf shows the openconnect binary now links to external libiconv/libintl (this is with CONFIG_BUILD_NLS=y).

```
 0x00000001 (NEEDED)                     Shared library: [libz.so.1]
 0x00000001 (NEEDED)                     Shared library: [libgnutls.so.30]
 0x00000001 (NEEDED)                     Shared library: [libxml2.so.2]
 0x00000001 (NEEDED)                     Shared library: [libintl.so.8]
 0x00000001 (NEEDED)                     Shared library: [libiconv.so.2]
 0x00000001 (NEEDED)                     Shared library: [libgcc_s.so.1]
 0x00000001 (NEEDED)                     Shared library: [libc.so]
 0x0000000f (RPATH)                      Library rpath: [/home/sk/tmp/openwrt/staging_dir/target-mips_24kc_musl/usr/lib/libintl-full/lib:/home/sk/tmp/openwrt/staging_dir/target-mips_24kc_musl/usr/lib/libiconv-full/lib]
```

Functionality-wise as an OpenWrt user I would expect it to follow the OpenWrt NLS configuration (in menuconfig). So if NLS is off you get iconv stub without NLS support and with NLS on you get everything via libiconv and libintl. openconnect using iconv from libc (musl/glibc) I would not expect, also because musl iconv is likely rather limited.

To sum things up, I'd suggest to use the change proposed by @mhei and add the configure switches on top.

Kind regards,
Seb

Comment (mhei) [reply net/openconnect/Makefile:39]: @nmav: 

> openconnect doesn't depend on iconv or libintl, so it shouldn't link to it. Since libxml2 depends in it,
> it should have done the right thing and added the right dependencies.
> That's what the rest of the libraries linked with like gnutls and openssl do.

I fully agree that in an ideal world, the API of libxml2 would be so abstracted that library users do need to bother whether libxml2 is iconv/intl capable or compiled against whatever other library. But this seems to be not the case...

I admit, that I missed the CONFIGURE_ARGS stuff since I did not run it by hand, I just reproduced the compile error and the absence of it with my proposal. I can easily add these lines and repush to this PR.
What do you think?

File: net/openconnect/Makefile [hunk 2]
```diff
@@ -34,7 +35,7 @@ endef
 define Package/openconnect
   SECTION:=net
   CATEGORY:=Network
-  DEPENDS:=+libxml2 +kmod-tun +resolveip +vpnc-scripts +OPENCONNECT_OPENSSL:libopenssl +OPENCONNECT_OPENSSL:p11-kit +OPENCONNECT_OPENSSL:libp11 +OPENCONNECT_GNUTLS:libgnutls +OPENCONNECT_GNUTLS:libtasn1 +OPENCONNECT_STOKEN:libstoken
+  DEPENDS:=+libxml2 +kmod-tun +resolveip +vpnc-scripts +OPENCONNECT_OPENSSL:libopenssl +OPENCONNECT_OPENSSL:p11-kit +OPENCONNECT_OPENSSL:libp11 +OPENCONNECT_GNUTLS:libgnutls +OPENCONNECT_GNUTLS:libtasn1 +OPENCONNECT_STOKEN:libstoken $(ICONV_DEPENDS) $(INTL_DEPENDS)
```
Comment (nmav) [inline net/openconnect/Makefile:38]: The ICONV depends should be added also on the OPENCONNECT_OPENSSL side

File: net/openconnect/Makefile (hunk 1 above)
Comment (nmav) [reply net/openconnect/Makefile:39]: But isn't the libxml2.pc file mentioning these libraries? If it is, there shouldn't be any need to explicitly link the dependencies as they would get these from pkg-config.

Comment (nmav) [reply net/openconnect/Makefile:39]: > Funnily enough openconnect does actually (try to) use iconv:
[...]

True, but there is little need to have openconnect with NLS in openwrt, at least no-one has asked for it.

> Functionality-wise as an OpenWrt user I would expect it to follow the OpenWrt NLS configuration (in menuconfig). So if NLS is off you get iconv stub without NLS support and with NLS on you get everything via libiconv and libintl. openconnect using iconv from libc (musl/glibc) I would not expect, also because musl iconv is likely rather limited.
> 
> To sum things up, I'd suggest to use the change proposed by @mhei and add the configure switches on top.

I still think that the libxml2.pc should contain the right libraries in order for openconnect to be able to link with libxml2. That way we can decide whether to bring libintl to openconnect independently, if needed, not just because libxml2 needs it.

**Comment by micmac1** [comment]
libxml2 doesn't use libintl. But openconnect wants libintl when it's compiled with NLS support. Sure, the dep on libiconv could be removed, but then if libiconv would get an ABI bump (and if ABI_VERSION was defined for libiconv), openconnect wouldn't be automatically recompiled. A lot of ifs, sure, I just would like to point that the explicit dep for openconnect makes sense, as openconnect itself will make use of these shared libs (to add another if, if NLS was selected). Hope this clarifies.

File: net/openconnect/Makefile (hunk 2 above)
Comment (mhei) [reply net/openconnect/Makefile:38]: I don't get it, sorry. Could you elaborate/rephrase?

Comment (nmav) [reply net/openconnect/Makefile:38]: Sorry,  that was a misunderstanding I got from the patch. What you have is correct.

File: net/openconnect/Makefile (hunk 1 above)
Comment (nmav) [reply net/openconnect/Makefile:39]: @mhei would it be possible to add libiconv into libxml2.pc? In that case doesn't it solve the compilation issue?

**Comment by nmav** [comment]
> libxml2 doesn't use libintl. But openconnect wants libintl when it's compiled with NLS support. Sure, the dep on libiconv could be removed, but then if libiconv would get an ABI bump (and if ABI_VERSION was defined for libiconv), openconnect wouldn't be automatically recompiled. A lot of ifs, sure, I just would like to point that the explicit dep for openconnect makes sense, as openconnect itself will make use of these shared libs (to add another if, if NLS was selected). Hope this clarifies.

I understand this part, but again I do not think we should add NLS support to openconnect just because libxml2 may be built with it. These are two independent things. libxml2 should still work with openconnect even if it supports iconv and openconnect doesn't have NLS.

**Comment by micmac1** [comment]
> > libxml2 doesn't use libintl. But openconnect wants libintl when it's compiled with NLS support. Sure, the dep on libiconv could be removed, but then if libiconv would get an ABI bump (and if ABI_VERSION was defined for libiconv), openconnect wouldn't be automatically recompiled. A lot of ifs, sure, I just would like to point that the explicit dep for openconnect makes sense, as openconnect itself will make use of these shared libs (to add another if, if NLS was selected). Hope this clarifies.
> 
> I understand this part, but again I do not think we should add NLS support to openconnect just because libxml2 may be built with it. These are two independent things. libxml2 should still work with openconnect even if it supports iconv and openconnect doesn't have NLS. So let's merge this, after we have fix libxml2 to add the right dependencies (iconv) in its .pc file.

This is not really an issue. Let me explain why. This commit does not add NLS support to openconnect. What it does is it adds the _option_ to have openconnect with NLS support. The user can activate NLS support via the global option "Compile with full language support". This is not default. That is to say the packages that the build bots compile are built with CONFIG_BUILD_NLS=n. In that case "--disable-nls" is passed to openconnect so it will not link to libintl. And the iconv library it finds is just a tiny stub archive, not libiconv-full.

libxml2 does not have nls support. So it will never link to libintl. But it does offer support for iconv. Previously this was simply disabled. But recently an issue was raised for it for a use case where libxml2 was involved in parsing RSS feeds and it failed to do it because it lacked the support for character encoding conversion (what iconv provides). So libxml2 package was updated to allow being built against iconv, either a small stub version with just some basics (CONFIG_BUILD_NLS=n) or libiconv-full (CONFIG_BUILD_NLS=y).

File: net/openconnect/Makefile (hunk 1 above)
Comment (micmac1) [reply net/openconnect/Makefile:39]: "-liconv" is already in the pc file. But to find the correct path to the lib that libxml2 was built against (be it iconv stub or libiconv-full) while building openconnect requires including nls.mk, because this include adds the correct paths depending on CONFIG_BUILD_NLS.

**Comment by nmav** [comment]
Ok, convinced :-) Thank you both for the argumentation!

**Review by nmav — APPROVED** [review]

**Comment by mhei** [comment]
Updated this PR with the two lines proposed by @micmac1.
